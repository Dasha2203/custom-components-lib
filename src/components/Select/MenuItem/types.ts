import { LiHTMLAttributes } from 'react';

export type MenuItemValue = string | number;

export type MenuItemProps = LiHTMLAttributes<HTMLLIElement> & {
  value: MenuItemValue;
  selected?: boolean;
  focused?: boolean;
};
