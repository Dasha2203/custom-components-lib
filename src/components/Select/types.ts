import { HTMLAttributes } from 'react';
import { MenuItemValue } from './MenuItem/types';

export type SelectProps = Omit<
  HTMLAttributes<HTMLDivElement>,
  'onChange' | 'value'
> & {
  label?: string;
  onChange?: (value: MenuItemValue) => void;
  value?: MenuItemValue;
  children?: React.ReactNode;
  disabled?: boolean;
};
