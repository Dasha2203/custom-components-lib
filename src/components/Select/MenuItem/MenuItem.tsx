import React from 'react';
import clsx from 'clsx';
import { MenuItemProps } from './types';
import * as classes from './style.module.scss';

const MenuItem = ({ children, selected, focused, ...props }: MenuItemProps) => {
  return (
    <li
      tabIndex={-1}
      role="menuitem"
      className={clsx(
        classes.menuItem,
        selected && classes.menuItem_selected,
        focused && classes.menuItem_focused,
        props.className
      )}
      {...props}
    >
      {children}
    </li>
  );
};

export default MenuItem;
