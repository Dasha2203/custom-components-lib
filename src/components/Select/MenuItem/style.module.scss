.menuItem {
  position: relative;
  padding: 6px 16px;
  box-sizing: border-box;
  display: flex;
  justify-content: flex-start;
  align-items: center;
  min-height: 48px;
  cursor: pointer;
  user-select: none;
  vertical-align: middle;
  color: inherit;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  letter-spacing: 0.00938em;
  white-space: nowrap;
  -webkit-tap-highlight-color: transparent;

  &:hover:not(.menuItem_selected, .menuItem_focused) {
    background-color: rgba(0, 0, 0, 0.04);
  }

  &_focused {
    background: rgba(224, 224, 224);
  }

  &_selected {
    background-color: rgba(25, 118, 210, 0.08);
  }
}
