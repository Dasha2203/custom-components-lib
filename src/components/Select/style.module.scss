.selectRoot {
  display: inline-flex;
  position: relative;
  flex-direction: column;
  min-width: 120px;
  outline: none;

  &:hover:not(.selectRoot_open, .selectRoot_disabled) {
    .select__fieldset {
      border-color: rgba(0, 0, 0, 0.87);
    }
  }

  &_active {
    .select__label {
      transform: translate(14px, -9px) scale(0.75);
    }

    .select__legend {
      max-width: 100%;
    }
  }
  &_disabled {
    .select {
      cursor: default;
    }
  }

  &_open {
    .select__icon {
      position: absolute;
      top: calc(50% - 0.5em);
      right: 7px;
      transform: rotate(180deg);
      width: 1em;
      height: 1em;
      display: inline-block;
      user-select: none;
      flex-shrink: 0;
      fill: currentcolor;
      font-size: 1.5rem;
      pointer-events: none;
      color: rgba(0, 0, 0, 0.54);
      transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1);
    }

    .select__fieldset {
      border-color: #1976d2;
      border-width: 2px;
    }

    .select__label {
      color: #1976d2;
    }
  }

  &__wrap {
    position: relative;
    box-sizing: border-box;
    display: inline-flex;
    align-items: center;
    font-weight: 400;
    font-size: 1rem;
    line-height: 1.4375em;
    letter-spacing: 0.00938em;
    color: rgba(0, 0, 0, 0.87);
    cursor: text;
    border-radius: 4px;
  }
}

.select {
  padding: 16.5px 14px;
  padding-right: 32px;
  display: block;
  width: 100%;
  min-width: 0;
  min-height: 1.4375em;
  letter-spacing: inherit;
  color: currentColor;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  outline: none;
  cursor: pointer;
  border-radius: 4px;
  -webkit-tap-highlight-color: transparent;

  &__fieldset {
    position: absolute;
    top: -5px;
    right: 0;
    bottom: 0;
    left: 0;
    padding: 0 8px;
    pointer-events: none;
    border-radius: inherit;
    border-style: solid;
    border-width: 1px;
    overflow: hidden;
    border-color: rgba(0, 0, 0, 0.23);
  }

  &__icon {
    user-select: none;
    position: absolute;
    top: calc(50% - 0.5em);
    right: 7px;
    display: inline-block;
    width: 1em;
    height: 1em;
    flex-shrink: 0;
    fill: currentColor;
    font-size: 1.5rem;
    color: rgba(0, 0, 0, 0.54);
    pointer-events: none;
    transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  }

  &__input {
    position: absolute;
    bottom: 0;
    left: 0;
    opacity: 0;
    pointer-events: none;
    width: 100%;
    box-sizing: border-box;
  }

  &__label {
    position: absolute;
    left: 0;
    top: 0;
    padding: 0;
    z-index: 1;
    display: block;
    max-width: calc(133% - 32px);
    color: rgba(0, 0, 0, 0.6);
    font-weight: 400;
    font-size: 1rem;
    line-height: 1.4375em;
    letter-spacing: 0.00938em;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    transform: translate(14px, 16px) scale(1);
    transition:
      color 200ms cubic-bezier(0, 0, 0.2, 1) 0ms,
      transform 200ms cubic-bezier(0, 0, 0.2, 1) 0ms,
      max-width 200ms cubic-bezier(0, 0, 0.2, 1) 0ms;
    user-select: none;
  }

  &__legend {
    padding: 0;
    display: block;
    max-width: 0.01px;
    height: 11px;
    overflow: hidden;
    font-size: 0.75em;
    visibility: hidden;
    transition: max-width 50ms cubic-bezier(0, 0, 0.2, 1) 0ms;
    white-space: nowrap;

    span {
      padding-inline: 5px;
      display: inline-block;
      opacity: 0;
      visibility: visible;
    }
  }
}
