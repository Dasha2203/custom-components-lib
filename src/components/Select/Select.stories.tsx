import React, { useState } from 'react';
import { Meta, StoryFn } from '@storybook/react';
import Select from './Select';
import { SelectProps } from './types';
import MenuItem from './MenuItem/MenuItem';

export default {
  title: 'Select',
  component: Select,
  tags: ['autodocs'],
  argTypes: {
    label: { control: 'text' },
    onChange: { action: 'changed' },
    value: { control: 'number' },
  },
} as Meta;

const Template: StoryFn<SelectProps> = (args) => {
  const [selectedValue, setSelectedValue] = useState(args.value || '');

  const handleChange = (value: string | number) => {
    setSelectedValue(value);
    args.onChange?.(value);
  };

  return (
    <Select {...args} value={selectedValue} onChange={handleChange}>
      <MenuItem value={10}>10</MenuItem>
      <MenuItem value={20}>20</MenuItem>
      <MenuItem value={30}>30</MenuItem>
    </Select>
  );
};

export const WithLabel = Template.bind({});
WithLabel.args = {
  label: 'Age',
  value: '',
  children: (
    <>
      <MenuItem value={10}>10</MenuItem>
      <MenuItem value={20}>20</MenuItem>
      <MenuItem value={30}>30</MenuItem>
    </>
  ),
};
