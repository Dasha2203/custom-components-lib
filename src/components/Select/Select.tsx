import React, { useEffect, useMemo, useRef, useState } from 'react';
import clsx from 'clsx';
import ArrowDropDownIcon from '@/icons/ArrowDropDownIcon';
import Dropdown from './Dropdown/Dropdown';
import { MenuItemValue, MenuItemProps } from './MenuItem/types';
import { SelectProps } from './types';
import * as classes from './style.module.scss';

const Select = ({
  value,
  label,
  children,
  disabled,
  onChange,
  ...props
}: SelectProps) => {
  const [isOpen, setIsOpen] = useState(false);
  const [isFocused, setIsFocused] = useState(false);
  const [focusedIndex, setFocusedIndex] = useState<number | null>(null);
  const selectRef = useRef<HTMLDivElement | null>(null);

  const handleFocus = () => {
    if (disabled) return;

    setIsFocused(true);
    setIsOpen(true);
  };

  const handleBlur = (event: React.FocusEvent) => {
    if (
      selectRef.current &&
      selectRef.current.contains(event.relatedTarget as Node)
    ) {
      return;
    }
    resetStates();
  };

  const resetStates = () => {
    setIsOpen(false);
    setIsFocused(false);
    setFocusedIndex(null);
  };

  const handleSelect = (value: MenuItemValue) => {
    onChange(value);
    resetStates();
  };

  const handleKeyDown = (event: React.KeyboardEvent<HTMLDivElement>) => {
    const options = React.Children.toArray(children);
    const optionsLength = React.Children.toArray(children).length;

    if (event.key === 'ArrowDown') {
      setFocusedIndex((prevIndex) =>
        prevIndex === null ? 0 : Math.min(prevIndex + 1, optionsLength - 1)
      );
    } else if (event.key === 'ArrowUp') {
      setFocusedIndex((prevIndex) =>
        prevIndex === null ? optionsLength - 1 : Math.max(prevIndex - 1, 0)
      );
    } else if (event.key === 'Enter' && focusedIndex !== null) {
      const el = options[focusedIndex];

      handleSelect(
        React.isValidElement<MenuItemProps>(el) ? el.props.value : ''
      );
    }
  };

  useEffect(() => {
    const controller = new AbortController();
    const signal = controller.signal;

    const handleClickOutside = (event: MouseEvent) => {
      if (
        selectRef.current &&
        !selectRef.current.contains(event.target as Node)
      ) {
        resetStates();
      }
    };

    document.addEventListener('click', handleClickOutside, { signal });

    return () => {
      controller.abort();
    };
  }, []);

  const selectedChild = useMemo(() => {
    const el = React.Children.toArray(children).find((child) => {
      if (React.isValidElement<MenuItemProps>(child)) {
        return child.props.value == value;
      }
      return false;
    });

    return React.isValidElement<MenuItemProps>(el) ? el.props.children : null;
  }, [value, children]);

  return (
    <div
      ref={selectRef}
      tabIndex={0}
      onFocus={handleFocus}
      onBlur={handleBlur}
      onKeyDown={handleKeyDown}
      className={clsx(
        classes.selectRoot,
        disabled && classes.selectRoot_disabled,
        isFocused && classes.selectRoot_open,
        (value || isFocused) && classes.selectRoot_active,
        props.className
      )}
      style={props.style}
    >
      <label className={classes.select__label}>{label}</label>
      <div className={classes.selectRoot__wrap}>
        <div
          tabIndex={0}
          role="combobox"
          aria-haspopup="listbox"
          aria-expanded={isOpen}
          aria-hidden={true}
          className={classes.select}
        >
          {selectedChild}
        </div>
        <input
          className={classes.select__input}
          aria-invalid="false"
          aria-hidden={true}
          tabIndex={-1}
          defaultValue={value}
        />
        <ArrowDropDownIcon className={classes.select__icon} />
        <fieldset aria-hidden="true" className={clsx(classes.select__fieldset)}>
          <legend className={classes.select__legend}>
            {label && <span>{label}</span>}
          </legend>
        </fieldset>
      </div>
      <Dropdown
        isOpen={isOpen}
        handleSelect={handleSelect}
        focused={focusedIndex}
        selected={value}
      >
        {children}
      </Dropdown>
    </div>
  );
};

export default Select;
