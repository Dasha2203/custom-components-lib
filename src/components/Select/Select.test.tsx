import '@testing-library/jest-dom';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import Select from './Select';
import MenuItem from './MenuItem/MenuItem';

describe('Testing Select component', () => {
  const selectId = 'select';
  const dropdownId = 'select-dropdown';

  test('Should render component', () => {
    render(<Select />);
    const select = screen.getByRole('combobox', { hidden: true });

    expect(select).toBeInTheDocument();
  });

  test('Should show label', () => {
    const label = 'Age';
    render(<Select label="Age" />);
    const labelEl = screen.getAllByText(label);

    expect(labelEl[0]).toBeInTheDocument();
  });

  test('Should show dropdown list after clicking by select element', () => {
    const label = 'Age';
    render(
      <Select label={label} value={10}>
        <MenuItem value={10}>10</MenuItem>
        <MenuItem value={20}>20</MenuItem>
        <MenuItem value={30}>30</MenuItem>
      </Select>
    );

    const select = screen.getByTestId(selectId);
    fireEvent.click(select);

    waitFor(() => {
      const dropdownEl = screen.getByTestId(dropdownId);

      expect(dropdownEl).toBeInTheDocument();
    });
  });

  test('Should select event work correctly', async () => {
    const handleChange = jest.fn();

    const { getByRole, findByTestId, rerender } = render(
      <Select label="Age" value={''} onChange={handleChange}>
        <MenuItem value={10}>10</MenuItem>
        <MenuItem value={20}>20</MenuItem>
        <MenuItem value={30}>30</MenuItem>
      </Select>
    );

    const select = getByRole('combobox', { hidden: true });
    fireEvent.focus(select);

    const dropdownEl = await findByTestId('select-dropdown');
    expect(dropdownEl).toBeInTheDocument();

    const option = dropdownEl.querySelector('li[value="10"]');
    fireEvent.click(option);

    expect(handleChange).toHaveBeenCalledWith(10);
    rerender(
      <Select label="Age" value={'10'} onChange={handleChange}>
        <MenuItem value={10}>10</MenuItem>
        <MenuItem value={20}>20</MenuItem>
        <MenuItem value={30}>30</MenuItem>
      </Select>
    );

    await waitFor(() => {
      expect(select).toHaveTextContent('10');
    });
  });

  test('Should not show dropdown list when passed disable property', () => {
    const label = 'Age';
    render(
      <Select label={label} value={10} disabled>
        <MenuItem value={10} data-testid={'option-10'}>
          10
        </MenuItem>
        <MenuItem value={20}>20</MenuItem>
        <MenuItem value={30}>30</MenuItem>
      </Select>
    );

    const select = screen.getByTestId('select');
    fireEvent.click(select);

    waitFor(() => {
      const dropdownEl = screen.queryByTestId('select-dropdown');

      expect(dropdownEl).not.toBeInTheDocument();
    });
  });
});

describe('Select snapshots', () => {
  const label = 'Age';
  const selectId = 'select';

  test('select element', async () => {
    render(
      <Select label={label}>
        <MenuItem value={10} data-testid={'option-10'}>
          10
        </MenuItem>
        <MenuItem value={20}>20</MenuItem>
        <MenuItem value={30}>30</MenuItem>
      </Select>
    );

    const select = screen.getByTestId(selectId);
    expect(select).toMatchSnapshot();
  });

  test('dropdown list', async () => {
    const { getByRole, findByTestId } = render(
      <Select label={label}>
        <MenuItem value={10} data-testid={'option-10'}>
          10
        </MenuItem>
        <MenuItem value={20}>20</MenuItem>
        <MenuItem value={30}>30</MenuItem>
      </Select>
    );

    const select = getByRole('combobox', { hidden: true });
    fireEvent.focus(select);

    const dropdownEl = await findByTestId('select-dropdown');
    expect(dropdownEl).toMatchSnapshot();
  });
});
