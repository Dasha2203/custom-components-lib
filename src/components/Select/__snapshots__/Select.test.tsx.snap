// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select snapshots dropdown list 1`] = `
<ul
  class="dropdown"
  data-testid="select-dropdown"
  role="listbox"
  tabindex="-1"
>
  <li
    class="menuItem"
    role="menuitem"
    tabindex="-1"
    value="10"
  >
    10
  </li>
  <li
    class="menuItem"
    role="menuitem"
    tabindex="-1"
    value="20"
  >
    20
  </li>
  <li
    class="menuItem"
    role="menuitem"
    tabindex="-1"
    value="30"
  >
    30
  </li>
</ul>
`;

exports[`Select snapshots select element 1`] = `
<div
  class="selectRoot"
  data-testid="select"
  tabindex="0"
>
  <label
    class="select__label"
  >
    Age
  </label>
  <div
    class="selectRoot__wrap"
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-hidden="true"
      class="select"
      role="combobox"
      tabindex="0"
    />
    <input
      aria-hidden="true"
      aria-invalid="false"
      class="select__input"
      tabindex="-1"
      value=""
    />
    <svg
      class="select__icon"
      data-testid="ArrowDropDownIcon"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
    <fieldset
      aria-hidden="true"
      class="select__fieldset"
    >
      <legend
        class="select__legend"
      >
        <span>
          Age
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
